drop-transfer/src/service.rs

```diff
@@ -111,6 +140,93 @@ impl Service {
         &self.state.storage
     }
 
+    async fn trigger_peer_incoming(&mut self, xfer: Arc<crate::IncomingTransfer>) {
+        debug!(self.logger, "trigger_peer_incoming() called: {:?}", xfer);
+
+        let _peer = xfer.peer();
+        let id = xfer.id();
+
+        if let Some(handle) = self.tasks.get(&id) {
+            if !handle.is_finished() {
+                return;
+            } else {
+                self.tasks.remove(&id);
+            }
+        }
+
+        let handle = check::spawn(
+            self.state.clone(),
+            xfer,
+            self.logger.clone(),
+            self.waiter.guard(),
+            self.stop.clone(),
+        );
+
+        self.tasks.insert(id, handle);
+    }
+
+    async fn trigger_peer_outgoing(&mut self, xfer: Arc<crate::OutgoingTransfer>) {
+        debug!(self.logger, "trigger_peer_outgoing() called: {:?}", xfer);
+
+        let id = xfer.id();
+        let _peer = xfer.peer();
+        if let Some(handle) = self.tasks.get(&id) {
+            if !handle.is_finished() {
+                return;
+            } else {
+                self.tasks.remove(&id);
+            }
+        }
+
+        let handle = ws::client::spawn(
+            self.state.clone(),
+            xfer,
+            self.logger.clone(),
+            self.waiter.guard(),
+            self.stop.clone(),
+        );
+
+        self.tasks.insert(id, handle);
+    }
+
+    pub async fn set_peer_state(&mut self, addr: &IpAddr, is_online: bool) {
```

Comment:
```suggestion
    pub async fn set_peer_state(&mut self, addr: IpAddr, is_online: bool) {
```
`IpAddr` is copyable